### DISCUSSION TITLE
Lab4 FIR從DMA接收到的資料錯誤

### DISCUSSION THREAD

**Discussion by alankuo04** [bol-edu/HLS-SOC-Discussions#215] [General]
如下圖，FIR會接收到兩筆0x00000010的資料而導致計算的結果錯誤。
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/68884383/ab45c98e-5dba-47c3-b56c-277371411bc3)

從FIR的輸出來看，確實是因為兩筆16的資料導致計算的結果錯誤。
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/68884383/983f73ee-d4bc-4891-9999-79561a3ebbe6)
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/68884383/0fe31414-2846-4286-8b7d-1b58194a5f76)

想請問有其他人遇到同樣的問題或是該如何解決嗎？ 謝謝！

**Comment by ruei7916** [comment]
我有遇到相同的問題
看起來是PL_AS有bug
![2024-05-06 03-49-49 的螢幕擷圖](https://github.com/bol-edu/HLS-SOC-Discussions/assets/56112123/795ada31-2f0c-4f14-8eb5-301a736b3c38)
波型圖可以看到userdma只有傳一次傳給AS，但AS轉傳給IS時卻變成兩次，最終導致fir收到重複的資料
從AS的code(https://github.com/bol-edu/fsic_fpga/blob/main/vivado/vvd_srcs/fpga/rtl/PL/sw_fpga.v) 可以看到as_is_tdata和as_is_tvalid單純只是up_as_tdata和up_as_tvalid經過一個flip-flop而已，而完全不管is_as_tready(line251, 364、line258, 385)

**Comment by boledulab** [comment]
我邀請原開發工程師一起討論這個問題。 時間： 5/6 4:00 pm， 下面是 Google Meet 連接。( Note: chagned from 4:00 pm to 5:30 pm )

FSIC Debugging 
Monday, May 6 · 5:30 – 6:30pm
Time zone: Asia/Taipei
Google Meet joining info
Video call link: https://meet.google.com/sjm-fqnf-iqn
Or dial: ‪(US) +1 475-277-0753‬ PIN: ‪759 569 578‬#
More phone numbers: https://tel.meet/sjm-fqnf-iqn?pin=6394497462121

**Reply by michael61112** [reply]
應該是打到講師說的這個issue
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/9332860/69c49ef2-a57f-47f0-be23-c840225078ca)

**Comment by boledulab** [comment]
Note: meeting time changed from 4:00 pm to 5:30 pm 

FSIC Debugging
Monday, May 6 · 5:30 – 6:30pm
Time zone: Asia/Taipei
Google Meet joining info
Video call link: https://meet.google.com/sjm-fqnf-iqn
Or dial: ‪(US) +1 475-277-0753‬ PIN: ‪759 569 578‬#
More phone numbers: https://tel.meet/sjm-fqnf-iqn?pin=6394497462121

**Comment by TonyHo722** [comment]
請大家幫我試一下這個 [commit](https://github.com/TonyHo722/fsic_fpga/commit/2eef856499064cfa926f20e28a16dfe874dd8904) in "0507_add_as_is" branch


```
$ git clone https://github.com/TonyHo722/fsic_fpga
$ cd fsic_fpga
$ git switch 0507_add_as_is
```

以下這三個file要更新
```
vivado/vvd_srcs/caravel_soc/rtl/user/io_serdes/rtl/io_serdes.v
vivado/vvd_srcs/caravel_soc/rtl/user/axis_switch/rtl/sw_caravel.v
vivado/vvd_srcs/fpga/rtl/PL/sw_fpga.v
```
目前試過可以解掉之前的pending issue - execution run_fsic_tb_sim 
- the result mismatch issue in updma_output_gold.log and updma_output.log
```
$ cd vivado
$ ./run_fsic_tb_sim
```
note: it is a draft version, I am still debugging this issue.

**Comment by TonyHo722** [comment]
Hi All,

Final code update is ready.
Please pull last code in https://github.com/bol-edu/fsic_fpga
Code update in this commit https://github.com/bol-edu/fsic_fpga/commit/09af236ad6658994672bbe5b187ef281533b49c0
```
vivado/vvd_srcs/caravel_soc/rtl/user/io_serdes/rtl/io_serdes.v
```
